Add the doc on user presence and audience (#6777)
microsoft/FluidFramework: docs/content/docs/build/audience.md

}
```

An `IMember` represents a single user identity.  `IMember` holds a list of `IConnection`s, which represent that audience member's active connections to the container.  Typically a user will only have one connection, but scenarios such as loading the container in multiple web contexts or on multiple computers will also result in as many connections.  An audience member will always have at least one connection.  Each user and each connection will both have a unique indentifier.

{{% callout tip %}}

> Connections can be short-lived and are not reused - a client that disconnects from the container and immediately reconnects will receive an entirely new connection.  The audience will reflect this as a member leaving and a member joining.

Author: i tried not to get too nitty between imember and iconnection throughout the doc, but it's the latter (and can also be the former if there is only one connection).  i'll clarify throughout

{{% /callout %}}

### Service-Specifics

> The `ServiceAudience` class represents the base audience implementation, and individual services are expected to extend this class for their needs.  Typically this is through extending `IMember` to provide richer user information and then extending `ServiceAudience` to use the `IMember` extension.  For `TinyliciousAudience`, this is the only change, and it defines a `TinyliciousMember` to add a user name.

Author: i think this one should stay as is because the suggestions are a functional change.  ServiceAudience is an abstract class so it must be extended.  extending IMember is the only required extension, but they can do other things too which i didn't think were worth mentioning here (like change client exclusions).

#### getMembers

The `getMembers` method returns a map of the audience's current members.  The map is keyed on user IDs (the same as the `IMember.userId` property), and values are the `IMember` for that user ID.  You can further query the individual `IMember`s for its client connections.

{{% callout tip %}}

> Because `ServiceAudience` exists to facilitate user presence scenarios, it may exclude certain client connections it doesn't consider useful for this purpose.  By default, this includes non-interactive clients such as the summarizer client (also see [Summarization](./will/this/page/ever/exist/idk)).

Author: I tried to frame this as a default behavior conforming to an intention that could be overridden (like theoretically someone could want summarizer clients and in their UI be like "hey everyone this user is responsible for summarizing").  If your more rigid use framing, probably something like "... don't represent "real users," such as the non-interactive [summarizer client]"?

{{% callout tip %}}

> Connection transitions can result in short timing windows where `getMyself` returns undefined.  This is because the current client connection will not have been added to the audience yet, so a matching connection ID cannot be found.  Similarly, offline scenarios may produce the same behavior.

Author: should be scenario specific, which is why i left it out.  i think this is an API issue we need to address though, so hopefully this callout will be short-lived

{{% /callout %}}

### Events

#### membersChanged

> The `membersChanged` event is emitted whenever an externally visible change to the audience members is made.  Listeners may call the `getMembers` method to get the new list of members.

Author: removed this since it gets into the specifics of the underlying container/audience and is only hinted at before in a callout

#### memberAdded

> The `memberAdded` event is emitted whenever an externally visible member or client connection is added to the audience.  The event also provides the connection client ID and the `IMember` object for this change.

Author: connection client id here refers just to the id, though that can used to find more information about the connection from the imember object.  will clarify

#### memberRemoved

The `memberRemoved` event is emitted whenver an externally visible member or client connection leaves the audience.  The event also provides the connection client ID and the `IMember` object for this change.

## Using Audience to Build Presence Features

### Data Management and Inter-User Communication

While the audience is the foundation for user presence features, the list of connected users does not provide a compelling experience on its own.  Your specific scenario will involve working with additional user data, which falls into a combination of the following three categories.

> #### Unshared Data

Reviewer: ```suggestion
#### Unshared data
```

I tend to think of this as "external and/or generated data". Is there a reason for the word "unshared?"

Author: it's to contrast with shared persisted and shared transient and form a set covering everything, because generated data could technically still need to be shared

> Not all presence scenarios will require sharing data among users - the data could be generated locally or fetched from an external service.  Consider the scenario where you want to display the connected users with a profile picture and a color border.  If you retrieve a user's profile picture from your service and assign each user a color based on a hash of their ID, you will have the desired data on other users without needing to communicate with them.

Author: i don't think it's necessary either

> Not all presence scenarios will require sharing data among users - the data could be generated locally or fetched from an external service.  Consider the scenario where you want to display the connected users with a profile picture and a color border.  If you retrieve a user's profile picture from your service and assign each user a color based on a hash of their ID, you will have the desired data on other users without needing to communicate with them.

Author: ```suggestion
In some cases, the user data could be generated locally or fetched from an external service. For example, consider a scenario where you want to display the connected users with a profile picture and a color border. If you retrieve a user's profile picture from your user metadata service and assign each user a color based on a hash of their user ID, you will have the desired data on other users without needing to communicate with them.
```

#### Shared Persisted Data

> Most presence scenarios will involve data that only a single user or client knows and needs to communicate to other audience members.  Some of those scenarios will require you to save data for each user for future sessions.  Consider if you want to display how long each user has spent in your application.  An active user's time should tick up, pause when they disconnect, and resume once they reconnect.  You may choose to use a `SharedMap` with a `SharedCounter` as the value onto which each user will increment their time spent every minute (also see [Introducing Distributed Data Structures](./dds.md)).  All other connected users will then receive changes to that map automatically.  Your UI can display data from the map for only users present in the audience.  A returning user can find themselves in the map and resume from the latest state.

Reviewer: If this is indeed "most scenarios" then it should be the first covered.

> Most presence scenarios will involve data that only a single user or client knows and needs to communicate to other audience members.  Some of those scenarios will require you to save data for each user for future sessions.  Consider if you want to display how long each user has spent in your application.  An active user's time should tick up, pause when they disconnect, and resume once they reconnect.  You may choose to use a `SharedMap` with a `SharedCounter` as the value onto which each user will increment their time spent every minute (also see [Introducing Distributed Data Structures](./dds.md)).  All other connected users will then receive changes to that map automatically.  Your UI can display data from the map for only users present in the audience.  A returning user can find themselves in the map and resume from the latest state.

Reviewer: ```suggestion
Most presence scenarios will involve data that only a single user or client knows and needs to communicate to other audience members.  Some of those scenarios will require you to save data for each user for future sessions.  For example, consider a scenario where you want to display how long each user has spent in your application.  An active user's time should increment while connected, pause when they disconnect, and resume once they reconnect.  This means that the time each user has spent must be persisted so it can survive disconnections.

One option is to use a `SharedMap` with a `SharedCounter` as the value onto which each user will increment their time spent every minute (also see [Introducing distributed data structures]({{< relref dds.md >}})).  All other connected users will then receive changes to that SharedMap automatically.  Your UI can display data from the map for only users present in the audience.  A returning user can find themselves in the map and resume from the latest state.
```

> #### Shared Transient Data

Reviewer: ```suggestion
#### Shared transient data
```

> Many presence scenarios involve data that are short-lived and do not need to be persisted.  Consider if you want to display where each user has selected in your UI.  Each user will need to tell other users their own information, and users do not care about past users.

Reviewer: ```suggestion
Many presence scenarios involve data that are short-lived and do not need to be persisted.  For example, consider a scenario where you want to display where each user has selected in your UI.  Each user will need to tell other users their own information -- where they clicked -- but the past data is irrelevant.
```

> From a data storage and communication perspective, you can accomplish this using DDSes in the same way as with the persisted data scenario.  However, using DDSes results in storage of data that are neither useful long term nor in contention among multiple users or clients.  [Signals](./where/does/this/article/go.md) are designed for sending transient data and would be more appropriate in this situation.  You can broadcast a signal containing your selection data to all connected users, and those users can store the data locally.  Newly connected users can request other connected users to send their selection data using another signal.  When a user disconnects, the local data is discarded.

Reviewer: ```suggestion
You can address this scenario using DDSes in the same way as with the persisted data scenario.  However, using DDSes results in storage of data that are neither useful long term nor in contention among multiple users or clients.  [Signals]({{< relref signals.md >}}) are designed for sending transient data and would be more appropriate in this situation.  Each user can broadcast a signal containing their selection data to all connected users, and those users can store the data locally.  Newly connected users can request other connected users to send their selection data using another signal.  When a user disconnects, the local data is discarded.
```
